fed/mosaic-application/src/main/java/org/eclipse/mosaic/fed/application/ambassador/simulation/electric/providers/ChargingStationIndex.java

```diff
@@ -0,0 +1,87 @@
+ *
+ * This program and the accompanying materials are made available under the
+ * terms of the Eclipse Public License 2.0 which is available at
+ * http://www.eclipse.org/legal/epl-2.0
+ *
+ * SPDX-License-Identifier: EPL-2.0
+ *
+ * Contact: mosaic@fokus.fraunhofer.de
+ */
+
+package org.eclipse.mosaic.fed.application.ambassador.simulation.electric.providers;
+
+import org.eclipse.mosaic.fed.application.ambassador.simulation.electric.objects.ChargingStationObject;
+import org.eclipse.mosaic.fed.application.ambassador.simulation.perception.index.objects.SpatialObjectAdapter;
+import org.eclipse.mosaic.lib.geo.GeoCircle;
+import org.eclipse.mosaic.lib.geo.GeoPoint;
+import org.eclipse.mosaic.lib.objects.electricity.ChargingStationData;
+import org.eclipse.mosaic.lib.spatial.KdTree;
+import org.eclipse.mosaic.lib.spatial.SpatialTreeTraverser;
+
+import java.util.ArrayList;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+
+public class ChargingStationIndex {
+
+    private final int bucketSize;
+
+    /**
+     * Stores {@link ChargingStationObject}s for fast removal and position update.
+     */
+    final Map<String, ChargingStationObject> indexedChargingStations = new HashMap<>();
+
+    private KdTree<ChargingStationObject> chargingStationTree;
+
+    private SpatialTreeTraverser.InRadius<ChargingStationObject> treeTraverser;
+
+    public ChargingStationIndex(int bucketSize) {
+        this.bucketSize = bucketSize;
+    }
+
+    /**
+     * Method called to initialize index after configuration has been read.
+     */
+    public void initialize() {
+        // initialization at first update
+    }
+
+    public void addChargingStation(String id, GeoPoint position) {
+        indexedChargingStations.computeIfAbsent(id, ChargingStationObject::new)
+                .setPosition(position.toCartesian());
+    }
+
+    /**
+     * Perform action before an update of the {@link ChargingStationIndex} takes place.
+     */
+    public void updateChargingStation(ChargingStationData chargingStationData) {
+        onChargingStationUpdate();
+        indexedChargingStations.get(chargingStationData.getName())
+                .setChargingStationData(chargingStationData);
+    }
+
+    public List<ChargingStationObject>getChargingStationsInCircle(GeoCircle circle) {
+        treeTraverser.setup(circle.getCenter().toVector3d(), circle.getRadius());
+        treeTraverser.traverse(chargingStationTree);
+        return treeTraverser.getResult();
+    }
+
+    public void onChargingStationUpdate() {
+        if (chargingStationTree == null) { // initialize before first update is called
+            List<ChargingStationObject> allChargingStations = new ArrayList<>(indexedChargingStations.values());
+            chargingStationTree = new KdTree<>(new SpatialObjectAdapter<>(), allChargingStations, bucketSize);
+            treeTraverser = new SpatialTreeTraverser.InRadius<>();
+        }
+    }
```

Comment:
Does this need to be public? I would make that private, rename it to `initSearchTree()` and move it right below `updateChargingStation` method where it is needed